pkg/object/cifs.go

```diff
@@ -0,0 +1,494 @@
+		_ = conn.session.Logoff()
+	}
+}
+
+func (c *cifsStore) withConn(f func(*cifsConn) error) error {
+	conn, err := c.getConnection()
+	if err != nil {
+		return err
+	}
+	err = f(conn)
+	c.releaseConnection(conn, err)
+	return err
+}
+
+func (c *cifsStore) Head(key string) (oj Object, err error) {
+	err = c.withConn(func(conn *cifsConn) error {
+		p := c.path(key)
+		fi, err := conn.share.Lstat(p)
+		if err != nil {
+			return err
+		}
+		isSymlink := fi.Mode()&os.ModeSymlink != 0
+		if isSymlink {
+			// SMB doesn't fully support symlinks like POSIX, but we'll try our best
+			fi, err = conn.share.Stat(p)
+			if err != nil {
+				return err
+			}
+		}
+		oj = c.fileInfo(key, fi, fi.Mode()&os.ModeSymlink != 0)
+		return nil
+	})
+	return oj, err
+}
+
+func (c *cifsStore) Get(key string, off, limit int64, getters ...AttrGetter) (io.ReadCloser, error) {
+	conn, err := c.getConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	p := c.path(key)
+	f, err := conn.share.Open(p)
+	if err != nil {
+		c.releaseConnection(conn, err)
+		return nil, err
+	}
+
+	if off > 0 {
+		_, err = f.Seek(off, 0)
+		if err != nil {
+			_ = f.Close()
+			c.releaseConnection(conn, err)
+			return nil, err
+		}
+	}
+
+	if limit > 0 {
+		return &smbReadCloser{
+			f:     f,
+			conn:  conn,
+			store: c,
+			limit: limit,
+		}, nil
+	}
+
+	return &smbReadCloser{
+		f:     f,
+		conn:  conn,
+		store: c,
+		limit: -1,
+	}, nil
+}
+
+// smbReadCloser is a wrapper for SMB file that handles connection release on close
+type smbReadCloser struct {
+	f     *smb2.File
+	conn  *cifsConn
+	store *cifsStore
+	limit int64
+	read  int64
+}
+
+func (r *smbReadCloser) Read(p []byte) (n int, err error) {
+	if r.limit >= 0 {
+		if r.read >= r.limit {
+			return 0, io.EOF
+		}
+		if int64(len(p)) > r.limit-r.read {
+			p = p[:r.limit-r.read]
+		}
+	}
+	n, err = r.f.Read(p)
+	r.read += int64(n)
+	return
+}
+
+func (r *smbReadCloser) Close() error {
+	err := r.f.Close()
+	r.store.releaseConnection(r.conn, err)
+	return err
+}
+
+func (c *cifsStore) Put(key string, in io.Reader, getters ...AttrGetter) (err error) {
+	return c.withConn(func(conn *cifsConn) error {
+		p := c.path(key)
+		if strings.HasSuffix(p, dirSuffix) {
+			// perm will not take effect, is not used
+			// ref: https://github.com/hirochachacha/go-smb2/blob/c8e61c7a5fa7bcd1143359f071f9425a9f4dda3f/client.go#L341-L370
+			return conn.share.MkdirAll(p, 0755)
+		}
+
+		// Ensure parent directories exist
+		dirPath := path.Dir(p)
+		if dirPath != "/" {
+			err = conn.share.MkdirAll(dirPath, 0755)
+			if err != nil {
+				return err
+			}
+		}
+
+		var tmp string
+		if PutInplace {
+			tmp = p
+		} else {
+			name := path.Base(p)
+			if len(name) > 200 {
+				name = name[:200]
+			}
+			tmp = path.Join(path.Dir(p), fmt.Sprintf(".%s.tmp.%d", name, rand.Int()))
+			defer func() {
+				if err != nil {
+					_ = conn.share.Remove(tmp)
+				}
+			}()
+		}
+
+		f, err := conn.share.Create(tmp)
+		if err != nil {
```

Comment:
If the parent directory does not exist, the parent directory should be created first before creating itself.